rfcs/20241015-result-accuracy.md

```diff
@@ -0,0 +1,103 @@
+accuracy and performance and also ensure more consistent numerical behaviors
+across different devices.
+
+### Caveats on targets
+
+The proposed result_accuracy attribute can only be supported for targets that
+have multiple implementations of the ops. For example, for XLA-CPU, the
+implementations of these ops are dependent on LLVM and each target may have a
+different implementation with a different level of accuracy. Thus, further
+analysis needs to be done to support this feature on CPUs. This analysis could
+be performed while building the compiler, assuming the compiler will only be
+used on a single type of CPU.
+
+## Proposed Specification
+
+### `result_accuracy`
+
+The users can specify the worst case numerical error they can tolerate in terms
+of absolute, relative and ULP (unit in last place) errors. If they don't care
+about the numerical accuracy, they can also choose the implementation using
+`mode`. We propose a new attribute `result_accuracy`. `result_accuracy` can be
+any combination of the following numerical tolerances `atol`, `rtol`, `ulps` or
+an enum of `HIGHEST`, `DEFAULT` or `TOLERANCE`. `TOLERANCE` enum is a default
+placeholder value for `mode` when the numerical tolerances are used. When using
+the numerical tolerances, at least one of atol, rtol or ulps should be
+specified. `HIGHEST` will give the most accurate implementation of the op
+and `DEFAULT` will give the fastest implementation with less accuracy.
+
+Name   | Type                  | Constraints
+------ | --------------------- | ---------------------------------
+`atol` | APFloat::IEEEdouble() | `atol >= 0`
+`rtol` | APFloat::IEEEdouble() | `rtol >= 0`
+`ulp`  | int64_t               | `ulp >= 0`
+`mode` | EnumAttr              | `HIGHEST`, `DEFAULT`, `TOLERANCE`
+
+```text
+New Attribute:
+  #stablehlo.result_accuracy<atol, rtol, ulps, mode=ResultAccuracyModeAttr>
+
+New Enum:
+  ResultAccuracyModeAttr ::= DEFAULT, HIGHEST, TOLERANCE
+```
+
+The default values are set as follows:
+
+```text
+#stablehlo.result_accuracy<atol, rtol, ulps, mode=EnumAttr>
+
+Case1: I want DEFAULT
+#stablehlo.result_accuracy<atol=0, rtol=0, ulps=0, mode=DEFAULT>
+
+Case2: I want HIGHEST
+#stablehlo.result_accuracy<atol=0, rtol=0, ulps=0, mode=HIGHEST>
+
+Case3: I want numerical tolerance X
+#stablehlo.result_accuracy<atol=X, rtol=X, ulps=X, mode=TOLERANCE>
+
+(C1) if mode != TOLERANCE: atol = rtol = ulps = 0
+```
+
+The numerical tolerances will be compared against the compiler errors according
+to the following inequality:
+
+`abs(expected(x)-actual(x)) <= max(abs(expected(x))*max(rtol, ulps*epsilon),
+atol)` for all x, where `epsilon` is the machine epsilon.
+
+The inequality will be checked against the errors of each implementation and the
+one that can satisfy the constraint will be returned. If multiple
+implementations satisfy the inequality, the faster implementation will be used.
+If none of the implementations can meet the requested tolerance, the compiler
```

Comment:
Naive question - Does this mean it's a runtime check that's inserted by the compiler?

Or there's some upfront guarantee / check that's promised by underlying implementations? My interpretation of this sentence makes it sound like the compiler has to first run the operation, then check that the result accuracy was correct and return if error.

Reply:
No runtime checks, the compiler will look check the request against its implementation errors at compile time.

Reply:
Ack mind making it explicit that it would be a compile time error?

Reply:
Thanks, added that it's a compile time error.